src/modm/platform/uart/stm32/freertos.hpp

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2009-2012, Fabian Greif
+ * Copyright (c) 2010, Martin Rosekeit
+ * Copyright (c) 2011, Georgi Grinshpun
+ * Copyright (c) 2011, 2013-2017, Niklas Hauser
+ * Copyright (c) 2012, Sascha Schade
+ * Copyright (c) 2013, 2016, Kevin Läufer
+ * Copyright (c) 2021, Raphael Lehmann
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_STM32_FREERTOS_UART_HPP
+#define MODM_STM32_FREERTOS_UART_HPP
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/queue.h>
+
+#include "uart_base.hpp"
+
+namespace modm::platform
+{
+
+/**
+ * Universal asynchronous receiver transmitter (implementaion based on FreeRTOS queues)
+ *
+ * @author      Kevin Laeufer
+ * @author      Niklas Hauser
+ * @author      Dima Barsky
+ * @ingroup     modm_platform_uart
+ */
+
+template <int SIZE>
+class FreeRtosBuffer
+{
+    StaticQueue_t queueStructure;
+    uint8_t storage[SIZE];
+public:
+    QueueHandle_t queue = xQueueCreateStatic(sizeof(storage), 1, storage, &queueStructure);
+};
+
+template <int SIZE>
+struct FreeRtosRxBuffer : public RxBuffer, FreeRtosBuffer<SIZE>{};
+
+template <int SIZE>
+struct FreeRtosTxBuffer : public TxBuffer, FreeRtosBuffer<SIZE>{};
+
+static inline
+void finalInterruptCallback(bool xHigherPriorityTaskWoken)
+{
+    portYIELD_FROM_ISR (xHigherPriorityTaskWoken);
+}
+
+template< class Hal, typename... Buffers>
+struct BufferedUart;
+
+template<int SIZE, class Hal, typename... Buffers>
+struct BufferedUart<Hal, FreeRtosTxBuffer<SIZE>, Buffers...>: public BufferedUart<Hal, Buffers...>
+{
+    using HasTxBuffer = std::true_type;
+    using Parent = BufferedUart<Hal, Buffers...>;
+    static_assert(! Parent::HasTxBuffer::value,
+                  "BufferedUart accepts at most one TxBuffer type");
+    inline static FreeRtosTxBuffer<SIZE> txBuffer;
+
+    static bool
+    write(uint8_t data)
+    {
+        if(transmitBufferSize() == 0 && Hal::isTransmitRegisterEmpty())
+            Hal::write(data);
+        else
+        {
+            if ( ! xQueueSend(txBuffer.queue, &data, 0) )
+                return false;
+            // Disable interrupts while enabling the transmit interrupt
+            atomic::Lock lock;
```

Comment:
I think this lock needs it's own scope { }, otherwise it locks already from the beginning of the `else {` statement.

Reply:
Are you sure about that? I always thought that local variables are initialised only when the execution reaches the point of their definition. I know that StackOverflow is not the ultimate authority on C++, but the guy who answered [this question](https://stackoverflow.com/questions/33987128) agrees with me. If you want I can find the relevant page on cppreference.com.

Reply:
The stack overflow answer is right. This is the [paragraph](https://eel.is/c++draft/stmt.dcl#2) from the standard:

> A block variable with automatic storage duration ([[basic.stc.auto]](https://eel.is/c++draft/basic.stc.auto)) is [active](https://eel.is/c++draft/stmt.dcl#def:variable,active) everywhere in the scope to which it belongs after its [init-declarator](https://eel.is/c++draft/dcl.decl.general#nt:init-declarator)[.](https://eel.is/c++draft/stmt.dcl#2.sentence-1)
Upon each transfer of control (including sequential execution of statements) within a function from point P to point Q, all block variables with automatic storage duration [...] that are active at Q but not at P are initialized in declaration order; unless all such variables have vacuous initialization ([[basic.life]](https://eel.is/c++draft/basic.life))

Reply:
Oh, interesting. TIL I guess?